src/estimagic/optimization/_trustregion_bounded_newton_quadratic.py

```diff
@@ -1,297 +1,365 @@
-            max_trustregion_radius,
-        ) = perform_steepest_descent_and_update_trustregion_radius(
+            radius_upper_bound,
+        ) = perform_gradient_descent(
             x_candidate,
-            f_candidate,
+            criterion_candidate,
             gradient_projected,
             hessian_inactive,
             model,
             lower_bounds,
             upper_bounds,
             active_bounds_info,
-            maxiter_steepest_descent,
+            maxiter_gradient_descent,
+            options_update_radius,
         )
 
-        if f_min_steepest_descent < f_candidate:
-            x_accepted = x_candidate_steepest_descent
-            f_accepted = f_min_steepest_descent
+        if f_min_gradient_descent < criterion_candidate:
+            criterion_candidate = f_min_gradient_descent
+
+            x_unbounded = (
+                x_candidate_gradient_descent
+                - step_size_gradient_descent * gradient_projected
+            )
+            x_candidate = apply_bounds_to_x_candidate(
+                x_unbounded, lower_bounds, upper_bounds
+            )
+
+            gradient_unprojected = evaluate_model_gradient(x_candidate, model)
+            active_bounds_info = get_information_on_active_bounds(
+                x_candidate,
+                gradient_unprojected,
+                lower_bounds,
+                upper_bounds,
+            )
 
-            gradient_candidate_steepest_descent = evaluate_model_gradient(
-                x_candidate_steepest_descent, model
+            gradient_projected = project_gradient_onto_feasible_set(
+                gradient_unprojected, active_bounds_info
+            )
+            hessian_inactive = find_hessian_submatrix_where_bounds_inactive(
+                model, active_bounds_info
             )
 
-            converged = check_for_convergence_steepest_descent(
-                x_candidate_steepest_descent,
-                f_min_steepest_descent,
-                gradient_candidate_steepest_descent,
+            converged, convergence_reason = check_for_convergence(
+                x_candidate,
+                criterion_candidate,
+                gradient_projected,
                 model,
                 lower_bounds,
                 upper_bounds,
-                gtol_abs,
-                gtol_rel,
-                gtol_scaled,
+                converged,
+                convergence_reason,
+                niter=None,
+                maxiter=None,
+                gtol_abs=gtol_abs,
+                gtol_rel=gtol_rel,
+                gtol_scaled=gtol_scaled,
             )
-        else:
-            x_accepted = x_candidate
-            f_accepted = f_candidate
 
-        trustregion_radius = max(trustregion_radius, max_trustregion_radius)
-        trustregion_radius = min(max(trustregion_radius, 1e-10), 1e10)
+        if not converged:
+            trustregion_radius = max(trustregion_radius, radius_upper_bound)
```

Comment:
This line looks redundant since you do clipping in the next line.

Reply:
`radius_upper_bound` and `options_update_radius["max_radius"]` are two different things. 
1) `radius_upper_bound` gets - potentially - updated during the gradient descent iterations.
2) `options_update_radius["max_radius"]` is a default option and does not get changed (maybe I could have been clearer with the naming here. I chose the `dict` type over a `namedtuple` to signal that the entries of `options_update_radius` do not get overwritten)

In line 141, we set `trustregion_radius` equal to `radius_upper_bound` in case we have not converged. Clipping is then done as a safeguarding step to ensure `trustregion_radius` lies within the default bounds, `"min_radius"` and `"max_radius"`.

Reply:
Ok!